### PR TITLE
Account Recovery: Resetting password through email

### PR DESCRIPTION
**Pull request by southp**
## Summary
This PR adds the step 5 -> step 6 of [the AR flow](https://cloudup.com/cC9EfchsqDg).
The following is a screencast of how it looks like. I'll explain why we need to modify the URL by hand in the Test Plan section.
demo-ar-email

Here is how it looks like when there is something wrong:

image

## Test Plan
Please note that this PR needs D5385 to work properly. By the time of testing, if that patch is not deployed yet, please remember to manually deploy it to your test environment.

### For successful case
1. Open an incognito window and make sure you are logged out.
1. Visit http://calypso.localhost:3000/account-recovery/
1. Choose a test account to proceed with the AR process. At the "Please choose a way to reset your password", choose either the primary email or the recovery email.
1. An email with a validation link should arrive your inbox soon. Click the link.
1. Note that it is pointing to `wordpress.com` since we are not detecting the request origin intentionally here. Please manually update the domain as `calypso.localhost:3000`.
1. After updating, you should see the "New Password" page, indicating the validation is succeeded.

### For error case
1. Follow the same steps as above.
1. When you have the validation link, try to mess up the query arguments, e.g. user, method, key. You should see the error message, then.

### CODE REVIEW

File: client/account-recovery/reset-password-email-validation/index.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+		return queryString.slice( 1 );
+	}
+
+	render() {
+		const { translate } = this.props;
+
+		return (
+			<Card>
+				<p>{ translate( 'Validating …' ) }</p>
```
Comment (a8ci18n) [inline client/account-recovery/reset-password-email-validation/index.js]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Validating' )` *ES Score: 28*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/account-recovery/ar-through-email)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/account-recovery/reset-code-validation/index.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	}
+
+	renderValidating = () => (
+		<p className="reset-code-validation__text-validating">{ this.props.translate( 'Validating' ) }</p>
+	)
+
+	renderErrorMsg = ( error ) => (
+		<p className="reset-code-validation__text-error">
+			{ error.message ? this.props.translate( "We've encountered an error: " ) + error.message
```
Comment (a8ci18n) [inline client/account-recovery/reset-code-validation/index.js]: Hi! I've found a possible matching string that has already been translated **23** times:
`translate( 'Autosave encountered an unexpected error' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/account-recovery/reset-code-validation/index.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			// arrive at wrong steps.
+			this.props.updatePasswordResetUserData( { user } );
+			this.props.setResetMethod( method );
+		}
+	}
+
+	renderErrorMsg = ( error ) => (
+		<Card className="reset-code-validation__content">
+			<p className="reset-code-validation__text-error">
+				{ error.message ? this.props.translate( "We've encountered an error: " ) + error.message
```
Comment (a8ci18n) [inline client/account-recovery/reset-code-validation/index.js]: Hi! I've found a possible matching string that has already been translated **23** times:
`translate( 'Autosave encountered an unexpected error' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by folletto** [comment]
Heya! I gave it a go :)
Nice work!

A couple of things:

1. Why is this accessible only to logged out users? I understand a logged in user would just go to Me and change it, but I'm wondering if we can't just make it available in general. What do you think?
2. Can we change the illustration to use `illustration-500.svg`? /cc @kjellr

**Comment by kristastevens** [comment]
If something goes wrong and the user sees an error screen, I think we should we offer something concrete on that error screen (if it's possible) to essentially chauffeur the user to the screen they need to be on to take the next step. I'd suggest putting a link on this page to the reset password screen so that we take them directly to the spot where they can choose another method. We could say:

"We've run into an unexpected error. We're sorry! 
Please try to reset your password again."

Where "reset your password" contains the link to the reset password page. 

Also, in case something's really gone wrong on our end and the user is really stuck (and maybe frustrated by this point), I think we should have a link to contact WordPress.com Support. There, we could say: 

"Having trouble? Contact WordPress.com Support and one of our Happiness Engineers will get you back on track." 

Where "Contact WordPress.com Support" links to `wordpress.com/help/contact/`

(I'm cribbing some [prior work](https://github.com/Automattic/automated-transfer/issues/179#issuecomment-298713463) on error messaging, which adds some consistency.)

**Comment by folletto** [comment]
> I think we should we offer something concrete on that error screen

100% agreed. Well pointed out!

**Comment by southp** [comment]
@kristastevens @folletto Thanks for your valuable feedbacks!

Now it comes with `illustration-500.svg` in two variations:

This is an "expected error" if the given reset link is expired or malformed somehow. The anchor will lead users to the beginning of the whole account recovery flow, `/account-recovery/`


This occurs when some unexpected errors happen, the same case as "something's really gone wrong on our end.".  

Again, the anchor is leading to `/account-recovery`. @kristastevens I didn't drop in `/help/contact` here because these screens are for logged-out users, so they don't actually have access to `/help/contact`. Does that make sense to you?

**Comment by southp** [comment]
@folletto Speaking of the logged-out restrictions, I would prefer to keep it for now. The REST APIs it uses are designed with users are logged out in mind, so I'm afraid removing the restrictions at this point would open a can of worms 😅 Of course, we can revisit this decision later if we found it useful for logged-in users.

How do you think?

**Comment by kristastevens** [comment]
First scenario looks fine, thanks for circling back, @southp!

In the case where: 

> This occurs when some unexpected errors happen, the same case as "something's really gone wrong on our end.".

The screen they're seeing is our fault for one reason or another -- something is down, something is broken and we're asking them to retry the same steps they just went through -- likely (if something is broken) to put them into a loop of "try again" and failure. This could get very frustrating for them very quickly. At this point I feel like we really need to give them a way to contact support to get help from a human. Is there a link that logged out used can use to get in touch with us?

**Comment by folletto** [comment]
> Speaking of the logged-out restrictions, I would prefer to keep it for now. [...]
> How do you think?

Ok for me, that's a good reason to wait. :)

> The anchor

The `EmptyContent` component provides a prop for having a button for next action. It's better to use that instead of linking the text.


> This could get very frustrating for them very quickly.

I agree with @kristastevens. I'm not 100% sure what's ideal to be linked here however... Can we maybe link to: https://wordpress.com/wp-login.php?action=recovery ? Who can we ask to know what's the best page there?

File: client/account-recovery/reset-code-validation/index.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+				title={ title }
+				line={ line }
+			/>
+		);
+	}
+
+	render = () => {
+		const { error } = this.props;
+
+		return error ? this.renderErrorMsg( error ) : null;
```
Comment (jordwest) [inline client/account-recovery/reset-code-validation/index.js]: `renderErrorMsg` and `getErrorMsg` might make good candidates to be broken out into separate pure functions/stateless functional components

File: client/account-recovery/reset-code-validation/index.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+			this.props.updatePasswordResetUserData( { user } );
+			this.props.setResetMethod( method );
+		}
+	}
+
+	getErrorMsg = ( errorIdentifier ) => {
+		const { translate } = this.props;
+
+		switch ( errorIdentifier ) {
+			case 'RestInvalidKeyError':
```
Comment (jordwest) [inline client/account-recovery/reset-code-validation/index.js]: Where does this string come from? Might be helpful to have it in `constants.js`

File: client/account-recovery/reset-code-validation/index.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		const queryString = get( window, 'location.search', null );
+
+		if ( ! queryString ) {
+			return '';
+		}
+
+		return queryString.slice( 1 );
+	}
+
+	parseQueryArgs = () => ( qs.parse( this.getQueryString() ) );
```
Comment (southp) [inline client/account-recovery/reset-code-validation/index.js]: Nope. It was an accident :D 
Updated in https://github.com/Automattic/wp-calypso/pull/13450/commits/ea57a2c7aed39a87a07bb823b81ce812148cf9ef#diff-fbc0e17344f63b1a5fb3da56568e5dc1L38

File: client/account-recovery/reset-code-validation/index.js (hunk 5 above)
Comment (southp) [inline client/account-recovery/reset-code-validation/index.js]: That was a good call. Extracted them in https://github.com/Automattic/wp-calypso/pull/13450/commits/ea57a2c7aed39a87a07bb823b81ce812148cf9ef#diff-e24dc6ce1bc226394765e25c090895b1R12 along with some other good candidates.

File: client/account-recovery/reset-code-validation/index.js (hunk 4 above)
Comment (southp) [inline client/account-recovery/reset-code-validation/index.js]: Good idea. I've extracted it as `AccountRecoveryErrorScreen` by https://github.com/Automattic/wp-calypso/pull/13450/commits/ea57a2c7aed39a87a07bb823b81ce812148cf9ef#diff-5efd6efbfa86c317ccdd23db6e3f00ac

**Comment by southp** [comment]
@folletto @kristastevens Thanks for bearing with me :D 
In ea57a2c, I've revamped it for mainly two things:

1. Replace anchors by action buttons.
1. The unexpected error screen is now a more serious error. The message is updated and the action button will lead the user to https://wordpress.com/wp-login.php?action=recovery after discussing with some HEs.

Now the screenshots:




How does that look to you? :)

**Comment by southp** [comment]
@jordwest Thanks for the helpful feedback as always :)

Although you have already approved this, to address your feedbacks I've refactored it a fair bit. Please help me revisit it when you have some spare time  :)

**Comment by folletto** [comment]
Design wise: 👍

**Comment by kristastevens** [comment]
Looks good! Thanks you, @southp!

**Pull request merged**